src/lein_monolith/plugin.clj

```diff
@@ -103,16 +103,38 @@
 (defn build-inherited-profiles
   "Returns a map from profile keys to inherited profile maps."
   [monolith subproject]
-  (reduce
-    (fn [acc [key config]]
-      (let [profile (some-> (choose-inheritance-source monolith config)
-                            (inherited-profile subproject config)
-                            (maybe-mark-leaky config))]
-        (if profile
-          (assoc acc key profile)
-          acc)))
-    nil
-    profile-config))
+  (when (:monolith/inherit subproject)
+    (reduce
+      (fn [acc [key config]]
+        (let [profile (some-> (choose-inheritance-source monolith config)
+                              (inherited-profile subproject config)
+                              (maybe-mark-leaky config))]
+          (if profile
+            (assoc acc key profile)
+            acc)))
+      nil
+      profile-config)))
+
+
+(defn build-dependency-set-profile
+  "Constructs a profile with managed dependencies from the subproject's dependency set.
+   Returns nil if the subproject does not use a dependency set."
+  [monolith subproject]
+  (when-let [dependency-set (:monolith/dependency-set subproject)]
+    (let [dependencies (or (get-in monolith [:monolith :dependency-sets dependency-set])
+                           (lein/abort (format "Unknown dependency set %s for project %s" dependency-set (:name subproject))))]
+      {:monolith/dependency-set
+       ^:leaky {:managed-dependencies dependencies}})))
+
+
+(defn build-profiles
+  "Constructs a map of profile keys to inherited profile maps and the dependency set profile map"
+  [monolith subproject]
+  (merge
+    ;; The dependency set profile should be the first profile, so that its managed dependencies
+    ;; take precedence.
+    (build-dependency-set-profile monolith subproject)
+    (build-inherited-profiles monolith subproject)))
```

Comment:
Merging maps together in Clojure _does not_ guarantee ordering! This happens to work at very small sizes because of the array-based implementation, but that is an implementation detail:
```clojure
=> (def a {:a 0, :b 1, :c 3})

=> (def b {:d 4, :e 5})

=> (keys a)
(:a :b :c)

=> (keys b)
(:d :e)

=> (keys (merge a b))
(:a :b :c :d :e)

=> (def c (into {} (map (juxt #(keyword (str (char (+ (int \d) %)))) #(+ % 3))) (range 10)))

=> c
{:m 12, :e 4, :l 11, :k 10, :g 6, :j 9, :h 7, :d 3, :f 5, :i 8}

=> (keys (merge a c))
(:m :e :l :k :g :c :j :h :b :d :f :i :a)
```
Instead, I think we need to control the order of the sequence provided to `project/merge-profiles` below.

Reply:
😓 Thanks, guess I'd taken too long of a break from Clojure